impeller/playground/imgui/imgui_impl_impeller.cc

```diff
@@ -150,14 +152,43 @@ void ImGui_ImplImpeller_RenderDrawData(ImDrawData* draw_data,
        draw_list_i++) {
     const ImDrawList* cmd_list = draw_data->CmdLists[draw_list_i];
 
+    // Convert ImGui's per-vertex data (`ImDrawVert`) into the per-vertex data
+    // required by the shader (`VS::PerVectexData`). The only difference is that
+    // `ImDrawVert` uses an `int` for the color and the impeller shader uses 4
+    // floats (since GLSL ES 1 doesn't support integer vertex attributes).
+    //
+    // Once the GLES 2 renderer lands, impellerc could be extended to map the 4
+    // color bytes to vec4s in the shader, i.e.:
+    //     glVertexAttribPointer(2, 4, GL_UNSIGNED_BYTE, false, 4, nullptr);
+    // ...as opposed to:
+    //     glVertexAttribPointer(2, 4, GL_FLOAT, false, 16, nullptr);
+    //
+    // impellerc could e.g. detect some kind of special naming convention hint
+    // from the attribute name to support this case:
+    //     in vec4 vertex_color__byte;
+    //
+    // Such a feature would remove the need to double-copy/convert this data.
+
+    std::vector<VS::PerVertexData> vtx_data;
+    vtx_data.reserve(cmd_list->VtxBuffer.size());
+    for (const auto& v : cmd_list->VtxBuffer) {
+      const impeller::Scalar kScale = 1.0f / 255.0f;
+      impeller::Vector4 color(((v.col >> 0) & 0xFF) * kScale,   //
+                              ((v.col >> 8) & 0xFF) * kScale,   //
+                              ((v.col >> 16) & 0xFF) * kScale,  //
+                              ((v.col >> 24) & 0xFF) * kScale);
+      vtx_data.push_back({impeller::Point(v.pos.x, v.pos.y),  //
+                          impeller::Point(v.uv.x, v.uv.y),    //
+                          color});
+    }
+
     auto draw_list_vtx_bytes =
-        static_cast<size_t>(cmd_list->VtxBuffer.size_in_bytes());
+        static_cast<size_t>(vtx_data.size() * sizeof(VS::PerVertexData));
```

Comment:
Probably (definitely) missing the forest for the trees here but `IMGUI_OVERRIDE_DRAWVERT_STRUCT_LAYOUT` will let you specify a custom layout (you can make it `PerVertexData`). That way, you can order the elements in place without an additional allocation.

You call on if you want to do this. Probably easier TBH and saves a copy. But this way works too.

Reply:
Cool! I think for this we would add some padding in the struct and then overwrite `col` + the extra space in place:
```
#define IMGUI_OVERRIDE_DRAWVERT_STRUCT_LAYOUT \
  struct ImDrawVert {                         \
    ImVec2 pos;                               \
    ImVec2 uv;                                \
    ImU32 col;                                \
    float[3] col_gba;                              \
  };
```

But I'm not certain it's ok to modify the `ImDrawList` data in-place?
I'm also slightly biased towards not requiring overrides in imconfig for this backend. If people use Impeller in the future for anything else, it'd be nice if they can just drop it in and it'll magically work.

Reply:
The comment for the macro seems to indicate that it is. But your approach is safer and the perf differences minimal.